iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+/// </summary>
+/// <param name="propertyValue">The unserialized property value.</param>
+/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
+/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
+/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
+public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
+{
+    PropertyValue = propertyValue;
+    AckCode = ackCode;
+    AckVersion = ackVersion;
+    AckDescription = ackDescription;
+}
+
+/// <summary>
+/// The unserialized property value.
+/// </summary>
+[JsonProperty("value")]
+public dynamic PropertyValue { get; set; }
+
+/// <summary>
+/// The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.
+/// </summary>
+[JsonProperty("ac")]
+public int AckCode { get; set; }
+
+/// <summary>
+/// The acknowledgement version, as supplied in the property update request.
+/// </summary>
+[JsonProperty("av")]
+public long AckVersion { get; set; }
+
+/// <summary>
+/// The acknowledgement description, an optional, human-readable message about the result of the property update.
+/// </summary>
+[JsonProperty("ad", DefaultValueHandling = DefaultValueHandling.Ignore)]
+public string AckDescription { get; set; }
+```
+</details>
+
+
+### Telemetry
+
+```csharp
+/// <summary>
+/// Send a single instance of telemetry.
+/// </summary>
+/// <param name="telemetryName">The name of the telemetry, as defined in the DTDL interface.</param>
+/// <param name="telemetryValue">The telemetry payload, in the format defined in the DTDL interface.</param>
+/// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(string telemetryName, dynamic telemetryValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Send a batched instance of telemetry.
+/// </summary>
+/// <param name="telemetryDictionary">The name and value telemetry pairs, as defined in the DTDL interface.</param>
+/// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(IDictionary<string, dynamic> telemetryDictionary, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Send a single instance of telemetry, formatted as per DTDL specifications:
+/// - the payload should be in the format: { "<telemetryName>" : "<telemetry_value>" }
```

Comment:
Why `telemetryName` but also `telemetry_value`? Different naming formats (camelCase vs snake_case).

Reply:
my bad, I'll update it all to be snake_case (I think it reads better in comments)